### PR TITLE
feat(wasi): add wasm32-wasi as a compile target

### PR DESCRIPTION
**Pull request by rioam2**
This PR adds support for cross-compiling to wasm32-wasi as a CMake target.

While WebAssembly has many ABI proposals (Emscripten and WASI being the most popular), I have evaluated WASI to be the most stable and future-proof, and therefore, have selected it as the preferred flavor of `wasm32`. The selection criteria came down to the following points:
- WASI has better support for various runtime environments than alternative ABIs (f.eks, it can be run in Golang, Rust, Javascript and the Command Line).
- WASI has a stronger ABI stability guarantee.

The drawback of using WASI is that the proposed specification is not as mature as Emscripten. For example, neither exceptions nor `setjmp`/`longjmp` have support as of now. Threading is supported, but requires a specific flavor of the wasi-sdk, which I have chosen not to implement in this PR for simplicity.

## Implementation

For this feature, I have bootstrapped [`wasi-sdk`](https://github.com/WebAssembly/wasi-sdk) as a CMake toolchain. The SDK is retrieved from the official Github release artifacts, and cached in the same cache directory as `vcpkg` is.

Using this toolchain, I have integrated it with `vcpkg` and added a CMake preset to build both the example CLI and module as WebAssembly binaries.

These binaries can be run inside any supported WASI runtime, such as [`wasmer.io`](https://wasmer.io/).
My preferred way of testing the CLI program is with wasmer's CLI tool:

```sh
# Build the WASM binary
cmake --workflow --preset=wasm32-wasi-clang-static

# Run the binary
curl https://get.wasmer.io -sSfL | sh
wasmer run build/wasm32-wasi-clang-static/src/cli/Debug/<module_name>_cli
```

This can be run on any host platform or architecture.

## Acknowledged shortcomings

As of now, I have not been able to compile Catch2 for `wasm32-wasi`. This can probably be done using a custom portfile for wasm32, but ends up being severely handicapped because of limited filesystem libc support. As such, the current CMake configuration disables tests when targeting `wasm32-wasi`. If the WASI standard matures to allow filesystem access, this can be revisited and added for CTest support on WebAssembly. Here is an example branch of how this would be done: https://github.com/rioam2/cpp-project-template/tree/feat-wasm32-wasi-catch2

Exceptions are not specified by the WASI specification nor supported when compiling for `wasm32-wasi`. This means code in your module cannot make use of `try..catch` blocks, or `throw`-ing exceptions. If these wish to be used on other targets, one might need to conditionally insert them using preprocessor checks. I have not yet investigated a better way for accomplishing this that does not degrade the experience for non `wasm32-wasi` targets.

### CODE REVIEW

File: cmake/presets/compilers/clang.json [hunk 1]
```diff
@@ -7,6 +7,9 @@
       "environment": {
         "CC": "clang",
         "CXX": "clang++"
+      },
+      "cacheVariables": {
+        "CMAKE_CXX_FLAGS": "-stdlib=libc++"
```
Comment (rioam2) [inline cmake/presets/compilers/clang.json:12]: This was needed to fix broken CI builds for clang on Ubuntu. Not strictly related to this PR.

File: cmake/presets/x64-linux-clang.json [hunk 2]
```diff
@@ -13,14 +13,18 @@
...
     },
     {
       "name": "x64-linux-clang-dynamic",
       "inherits": "x64-linux-clang-static",
       "displayName": "x64 linux clang dynamic libs",
       "cacheVariables": {
-        "BUILD_SHARED_LIBS": true
+        "BUILD_SHARED_LIBS": true,
+        "CMAKE_CXX_FLAGS": "-stdlib=libc++"
       }
```
Comment (rioam2) [inline cmake/presets/x64-linux-clang.json:28]: This was needed to fix broken CI builds for clang on Ubuntu. Not strictly related to this PR.

**Comment by rioam2** [comment]
@msvetkin I've now added an overlay port to get the `fmt` package working on `wasm32-wasi`. 
Now only `catch2` does not compile tests for WebAssembly, which is expected since the proposal for WASI has not specified any filesystem access implementations. This will need to wait for the specification to mature more.

File: .github/actions/wasi/base/action.yml [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+      shell: bash
+      run: |
+        sudo apt-get update
+        sudo apt-get install -y ninja-build wget
+    - name: Install CMake
+      shell: bash
+      run: |
+        wget -nv -O ./cmake_install.sh https://github.com/Kitware/CMake/releases/download/v3.28.0-rc5/cmake-3.28.0-rc5-linux-x86_64.sh
+        sudo bash ./cmake_install.sh --skip-license --prefix=/usr --exclude-subdir
+        rm -f ./cmake_install.sh
```
Comment (msvetkin) [inline .github/actions/wasi/base/action.yml:16]: Could we use cmake which provide by ubuntu? 
If yes, then we can just re-use action from linux/base

File: .github/actions/wasi/clang/action.yml [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+  using: "composite"
+  steps:
+    - name: Install clang/llvm 17
+      shell: bash
+      run: |
+        sudo apt-add-repository 'deb https://apt.llvm.org/jammy llvm-toolchain-jammy-17 main'
+        sudo wget -qO /etc/apt/trusted.gpg.d/llvm.asc https://apt.llvm.org/llvm-snapshot.gpg.key
+        sudo apt-get update
+        sudo apt-get install -y -t llvm-toolchain-jammy-17 \
+            clang-17 llvm-17 lld-17 lldb-17 libc++-17-dev \
```
Comment (msvetkin) [inline .github/actions/wasi/clang/action.yml:13]: Is it clang-17 minimum supported version?

File: CMakeLists.txt [hunk 5]
```diff
@@ -9,5 +9,7 @@ list(APPEND CMAKE_MODULE_PATH "${PROJECT_SOURCE_DIR}/cmake")
 
 add_subdirectory(src)
 
-enable_testing()
-add_subdirectory(tests)
+if (NOT WASI) 
+    enable_testing()
+    add_subdirectory(tests)
```
Comment (msvetkin) [inline CMakeLists.txt:14]: What is the issue with tests?

File: cmake/presets/x64-linux-clang.json [hunk 6]
```diff
@@ -13,14 +13,18 @@
         "linux",
         "clang"
       ],
-      "displayName": "x64 linux clang static libs"
+      "displayName": "x64 linux clang static libs",
+      "cacheVariables": {
+        "CMAKE_CXX_FLAGS": "-stdlib=libc++"
```
Comment (msvetkin) [inline cmake/presets/x64-linux-clang.json:18]: This flags should be moved to `src/cmake/set_cpp_pt_target_properties.cmake`.
presets are used only for developers and consumers use `src`.

File: cmake/vcpkg/ports/wasm32-wasi/fmt/README.md [hunk 7]
```diff
@@ -0,0 +1 @@
+This overlay port is a direct copy from https://github.com/microsoft/vcpkg/tree/master/ports/fmt with modifications to allow compilation on wasm32-wasi as a target by removing POSIX filesystem calls.
```
Comment (msvetkin) [inline cmake/vcpkg/ports/wasm32-wasi/fmt/README.md:1]: According to fmt github repo it supports build with emscripten.
Can you share a compilation error?

File: .github/actions/wasi/wasm32/action.yml [hunk 8]
```diff
@@ -0,0 +1,12 @@
...
+
+runs:
+  using: "composite"
+  steps:
+    - name: Download wasm32 libclang runtime
+      shell: bash
+      run: |
+        sudo mkdir -p "$(clang --print-resource-dir)"
+        sudo wget -c \
+          https://github.com/WebAssembly/wasi-sdk/releases/download/wasi-sdk-21/libclang_rt.builtins-wasm32-wasi-21.0.tar.gz \
```
Comment (msvetkin) [inline .github/actions/wasi/wasm32/action.yml:11]: Don't we download it via `wasi-sdk_bootstrap.cmake`?

File: src/cmake/add_cpp_pt_module.cmake [hunk 9]
```diff
@@ -11,10 +11,23 @@ include(GNUInstallDirs)
...
+  set(export_file "${export_file_dir}/${module_name}/export.hpp")
   generate_export_header(${module_target}
-    EXPORT_FILE_NAME "${export_file_dir}/${module_name}/export.hpp"
+    EXPORT_FILE_NAME ${export_file}
   )
 
+  if (WASI)
+    file(APPEND ${export_file} "\
+      \n#ifndef WASI_EXPORT\
+      \n#define WASI_EXPORT(name) extern \"C\" __attribute__((export_name(name)))\
```
Comment (msvetkin) [inline src/cmake/add_cpp_pt_module.cmake:22]: What is the issue with generated export macros?

File: .github/actions/wasi/base/action.yml (hunk 3 above)
Comment (rioam2) [reply .github/actions/wasi/base/action.yml:16]: Good mention. I thought a newer version of CMake was needed, but it was not. 
Using the ubuntu-provided cmake now :)

File: cmake/vcpkg/ports/wasm32-wasi/fmt/README.md (hunk 7 above)
Comment (rioam2) [reply cmake/vcpkg/ports/wasm32-wasi/fmt/README.md:1]: In short: `emscripten != wasi`. This PR allows compiling with WASI, not Emscripten.

WebAssembly has two popular ABI flavors: Emscripten and WASI.
Emscripten is more feature-forward at the cost of being non-standard and only supporting Javascript as a runtime host.

WASI is more limited in capabilities but is standardized and can be run in many more runtimes (Javascript, Golang, Rust, CLI, ...).

Here's the error without the overlay port:

```Change Dir: '/Users/riomartinez/.cache/vcpkg/projects/crdt/buildtrees/fmt/wasm32-wasi-dbg'

Run Build Command(s): /opt/homebrew/bin/ninja -v -v -j11 install
[1/4] /opt/homebrew/opt/llvm/bin/clang++ --target=wasm32-wasi --sysroot=/Users/riomartinez/.cache/wasi-sdk/21.0/wasi-sysroot  -I/Users/riomartinez/.cache/vcpkg/projects/crdt/buildtrees/fmt/src/10.0.0-cbbdb219f9.clean/include -g -fvisibility=hidden -fvisibility-inlines-hidden -stdlib=libc++ -fno-exceptions -D_WASI_EMULATED_SIGNAL -MD -MT CMakeFiles/fmt.dir/src/os.cc.obj -MF CMakeFiles/fmt.dir/src/os.cc.obj.d -o CMakeFiles/fmt.dir/src/os.cc.obj -c /Users/riomartinez/.cache/vcpkg/projects/crdt/buildtrees/fmt/src/10.0.0-cbbdb219f9.clean/src/os.cc
FAILED: CMakeFiles/fmt.dir/src/os.cc.obj 
/opt/homebrew/opt/llvm/bin/clang++ --target=wasm32-wasi --sysroot=/Users/riomartinez/.cache/wasi-sdk/21.0/wasi-sysroot  -I/Users/riomartinez/.cache/vcpkg/projects/crdt/buildtrees/fmt/src/10.0.0-cbbdb219f9.clean/include -g -fvisibility=hidden -fvisibility-inlines-hidden -stdlib=libc++ -fno-exceptions -D_WASI_EMULATED_SIGNAL -MD -MT CMakeFiles/fmt.dir/src/os.cc.obj -MF CMakeFiles/fmt.dir/src/os.cc.obj.d -o CMakeFiles/fmt.dir/src/os.cc.obj -c /Users/riomartinez/.cache/vcpkg/projects/crdt/buildtrees/fmt/src/10.0.0-cbbdb219f9.clean/src/os.cc
/Users/riomartinez/.cache/vcpkg/projects/crdt/buildtrees/fmt/src/10.0.0-cbbdb219f9.clean/src/os.cc:273:31: error: no member named 'dup' in the global namespace
  273 |   int new_fd = FMT_POSIX_CALL(dup(fd));
      |                ~~~~~~~~~~~~~~~^~~~~~~~
/Users/riomartinez/.cache/vcpkg/projects/crdt/buildtrees/fmt/src/10.0.0-cbbdb219f9.clean/include/fmt/os.h:56:36: note: expanded from macro 'FMT_POSIX_CALL'
   56 | #    define FMT_POSIX_CALL(call) ::call
      |                                  ~~^
/Users/riomartinez/.cache/vcpkg/projects/crdt/buildtrees/fmt/src/10.0.0-cbbdb219f9.clean/src/os.cc:282:36: error: no member named 'dup2' in the global namespace
  282 |   FMT_RETRY(result, FMT_POSIX_CALL(dup2(fd_, fd)));
      |                     ~~~~~~~~~~~~~~~^~~~~~~~~~~~~~
/Users/riomartinez/.cache/vcpkg/projects/crdt/buildtrees/fmt/src/10.0.0-cbbdb219f9.clean/include/fmt/os.h:56:36: note: expanded from macro 'FMT_POSIX_CALL'
   56 | #    define FMT_POSIX_CALL(call) ::call
      |                                  ~~^
/Users/riomartinez/.cache/vcpkg/projects/crdt/buildtrees/fmt/src/10.0.0-cbbdb219f9.clean/include/fmt/os.h:71:61: note: expanded from macro 'FMT_RETRY'
   71 | #define FMT_RETRY(result, expression) FMT_RETRY_VAL(result, expression, -1)
      |                                                             ^~~~~~~~~~
/Users/riomartinez/.cache/vcpkg/projects/crdt/buildtrees/fmt/src/10.0.0-cbbdb219f9.clean/include/fmt/os.h:65:19: note: expanded from macro 'FMT_RETRY_VAL'
   65 |       (result) = (expression);                            \
      |                   ^~~~~~~~~~
/Users/riomartinez/.cache/vcpkg/projects/crdt/buildtrees/fmt/src/10.0.0-cbbdb219f9.clean/src/os.cc:292:36: error: no member named 'dup2' in the global namespace
  292 |   FMT_RETRY(result, FMT_POSIX_CALL(dup2(fd_, fd)));
      |                     ~~~~~~~~~~~~~~~^~~~~~~~~~~~~~
/Users/riomartinez/.cache/vcpkg/projects/crdt/buildtrees/fmt/src/10.0.0-cbbdb219f9.clean/include/fmt/os.h:56:36: note: expanded from macro 'FMT_POSIX_CALL'
   56 | #    define FMT_POSIX_CALL(call) ::call
      |                                  ~~^
/Users/riomartinez/.cache/vcpkg/projects/crdt/buildtrees/fmt/src/10.0.0-cbbdb219f9.clean/include/fmt/os.h:71:61: note: expanded from macro 'FMT_RETRY'
   71 | #define FMT_RETRY(result, expression) FMT_RETRY_VAL(result, expression, -1)
      |                                                             ^~~~~~~~~~
/Users/riomartinez/.cache/vcpkg/projects/crdt/buildtrees/fmt/src/10.0.0-cbbdb219f9.clean/include/fmt/os.h:65:19: note: expanded from macro 'FMT_RETRY_VAL'
   65 |       (result) = (expression);                            \
      |                   ^~~~~~~~~~
/Users/riomartinez/.cache/vcpkg/projects/crdt/buildtrees/fmt/src/10.0.0-cbbdb219f9.clean/src/os.cc:309:31: error: no member named 'pipe' in the global namespace
  309 |   int result = FMT_POSIX_CALL(pipe(fds));
      |                ~~~~~~~~~~~~~~~^~~~~~~~~~
/Users/riomartinez/.cache/vcpkg/projects/crdt/buildtrees/fmt/src/10.0.0-cbbdb219f9.clean/include/fmt/os.h:56:36: note: expanded from macro 'FMT_POSIX_CALL'
   56 | #    define FMT_POSIX_CALL(call) ::call
      |                                  ~~^
4 errors generated.
[2/4] /opt/homebrew/opt/llvm/bin/clang++ --target=wasm32-wasi --sysroot=/Users/riomartinez/.cache/wasi-sdk/21.0/wasi-sysroot  -I/Users/riomartinez/.cache/vcpkg/projects/crdt/buildtrees/fmt/src/10.0.0-cbbdb219f9.clean/include -g -fvisibility=hidden -fvisibility-inlines-hidden -stdlib=libc++ -fno-exceptions -D_WASI_EMULATED_SIGNAL -MD -MT CMakeFiles/fmt.dir/src/format.cc.obj -MF CMakeFiles/fmt.dir/src/format.cc.obj.d -o CMakeFiles/fmt.dir/src/format.cc.obj -c /Users/riomartinez/.cache/vcpkg/projects/crdt/buildtrees/fmt/src/10.0.0-cbbdb219f9.clean/src/format.cc
ninja: build stopped: subcommand failed.

```

File: .github/actions/wasi/wasm32/action.yml (hunk 8 above)
Comment (rioam2) [reply .github/actions/wasi/wasm32/action.yml]: > Don't we download it via `wasi-sdk_bootstrap.cmake`?

Yep! Thanks for the notice. I've removed this unnecessary step 👍

File: cmake/vcpkg/ports/wasm32-wasi/fmt/README.md (hunk 7 above)
Comment (rioam2) [reply cmake/vcpkg/ports/wasm32-wasi/fmt/README.md:1]: For reference, here is the code from `wasi-libc` which can be seen as a source-of-truth for what system calls are supported in WASI: https://github.com/WebAssembly/wasi-libc/blob/6593687e25f07526c4b92a20fe5ddf507599d5b3/libc-top-half/musl/include/unistd.h#L54

As can be seen from the link, `dup`, `dup2` and `pipe` are not supported by WASI.

**Comment by rioam2** [comment]
Superseded by #13

**Pull request closed without merging**